File: 258_Add_Digits.java
/*
https://leetcode.com/problems/add-digits/
Given a non-negative integer num, repeatedly add all its digits until the result has only one digit.

Digital root.  Alternative closed form expression is (n-1)%9 + 1.
Time - O(log n)
Space - O(1)
*/

public class Solution {
    public int addDigits(int num) {

        while (num > 9) {           // more than one digit

            int digitSum = 0;
            while (num > 0) {       // sum all digits
                digitSum += num % 10;
                num /= 10;
            }
            num = digitSum;

        }
        return num;
    }
}

File: 258_Add_Digits.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/add-digits/
# Given a non-negative integer num, repeatedly add all its digits until the result has only one digit.

# Repeatedly sum digitss until < 10.
# Time - O(log n)
# Space - O(1)

class Solution(object):
    def addDigits(self, num):
        """
        :type num: int
        :rtype: int
        """
        while num > 9:
            num = sum([int(c) for c in str(num)])
        return num